## Final Project Submission

Please fill out:
* Student name: John Elvis
* Student pace: part time 
* Scheduled project review date/time: 
* Instructor name: 
* Blog post URL:


## Overview

This project employs exploratory data analysis to extract valuable information for Microsoft, aiming to establish a new movie studio. By analyzing datasets from IMDB and Box Office Mojo, the project provides recommendations regarding the film genres that Microsoft should consider exploring. 

The success and popularity of each genre are evaluated using two key metrics: the total gross incomes and the critical response. The results suggest that Microsoft would benefit from concentrating on genres such as Sci-Fi, adventure, animation, and action, as these genres have demonstrated higher total gross income.

## Business Problem

Microsoft intends to join the league of prominent companies by venturing into the production of unique video content. To accomplish this, they plan to establish a movie studio. However, their lack of expertise in the field poses a challenge. This analysis primarily focuses on examining the prevailing trends in successful movies at the box office. Through an exploratory data analysis, valuable insights are generated, which can be used by the head of Microsoft's new movie studio to make informed decisions regarding the genres and themes of films to produce.

Questions :<br>
Which genre of movies have the highest gross income?<br>
Which genre of movies have a higher average rating?<br>
What is the relationship between the average rating and the gross income?<br>
What is the domestic gross and foreign gross income by start year of the movies?<br>

### Importing Modules and Relevant Datasets

In [39]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt


In [40]:
#load the datasets
movie_gross = pd.read_csv('zippedData/bom.movie_gross.csv.gz')
imdb_title_basics = pd.read_csv('zippedData/imdb.title.basics.csv.gz')
imdb_ratings = pd.read_csv('zippedData/imdb.title.ratings.csv.gz')

### Dataset understanding

We will define a function `df_inpsect` to help inspect our datasets.<br>
It returns : <br>
- the first 5 entries of the dataset <br> 
- dataset information <br>
- total number of null values per column and <br> 



In [47]:
def df_inspect(df):
    """
    Takes a DataFrame as input, and displays the head, info, and sum of all null values for each column in that DataFrame
    """
    return (display(df.head()), 
            display(df.info()), 
            display(df.isna().sum()))


#### Movie_gross dataframe

In [48]:
#inspect movie_gross dataframe
df_inspect(movie_gross)

title studio  domestic_gross  \
0                                  Toy Story 3     BV     415000000.0   
1                   Alice in Wonderland (2010)     BV     334200000.0   
2  Harry Potter and the Deathly Hallows Part 1     WB     296000000.0   
3                                    Inception     WB     292600000.0   
4                          Shrek Forever After   P/DW     238700000.0   

  foreign_gross  year  
0     652000000  2010  
1     691300000  2010  
2     664300000  2010  
3     535700000  2010  
4     513900000  2010

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3387 entries, 0 to 3386
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           3387 non-null   object 
 1   studio          3382 non-null   object 
 2   domestic_gross  3359 non-null   float64
 3   foreign_gross   2037 non-null   object 
 4   year            3387 non-null   int64  
dtypes: float64(1), int64(1), object(3)
memory usage: 132.4+ KB


None

title                0
studio               5
domestic_gross      28
foreign_gross     1350
year                 0
dtype: int64

(None, None, None)

From the above output, DataFrame `bom_movie` contains 3387 rows and 5 columns with the following information about movies:
> - title : The title of the movie <br>
> - studio : The studio that produced the movie<br>
> - domestic_gross : The domestic gross revenue of the movie in <br>
> - foreign_gross : The foreign gross revenue of the movie in dollars<br>
> - year : The year in which the movie was released<br>

In [50]:
df_inspect(imdb_title_basics)

tconst                    primary_title              original_title  \
0  tt0063540                        Sunghursh                   Sunghursh   
1  tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
2  tt0069049       The Other Side of the Wind  The Other Side of the Wind   
3  tt0069204                  Sabse Bada Sukh             Sabse Bada Sukh   
4  tt0100275         The Wandering Soap Opera       La Telenovela Errante   

   start_year  runtime_minutes                genres  
0        2013            175.0    Action,Crime,Drama  
1        2019            114.0       Biography,Drama  
2        2018            122.0                 Drama  
3        2018              NaN          Comedy,Drama  
4        2017             80.0  Comedy,Drama,Fantasy

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 146144 entries, 0 to 146143
Data columns (total 6 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   tconst           146144 non-null  object 
 1   primary_title    146144 non-null  object 
 2   original_title   146123 non-null  object 
 3   start_year       146144 non-null  int64  
 4   runtime_minutes  114405 non-null  float64
 5   genres           140736 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 6.7+ MB


None

tconst                 0
primary_title          0
original_title        21
start_year             0
runtime_minutes    31739
genres              5408
dtype: int64

(None, None, None)

From the above output, DataFrame `imdb_title_basics` contains 146144 rows and 6 columns 

In [44]:
df_inspect(imdb_ratings)

tconst  averagerating  numvotes
0  tt10356526            8.3        31
1  tt10384606            8.9       559
2   tt1042974            6.4        20
3   tt1043726            4.2     50352
4   tt1060240            6.5        21

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73856 entries, 0 to 73855
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   tconst         73856 non-null  object 
 1   averagerating  73856 non-null  float64
 2   numvotes       73856 non-null  int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 1.7+ MB


None

tconst           0
averagerating    0
numvotes         0
dtype: int64

None None None


From the above output, DataFrame `imdb_title_basics` contains 73856 rows and 3 columns 

### Data Cleaning

The next step is to clean our data so that we can deal with quality data to enhance analysis.
Steps:
 - Check for missing values
 - Check for duplicate values
 - Amend data types 
 - Fix error values

#### Check for missing values

We create a simple loop to create a list of the entries with missing values for each column of a dataset.

In [54]:
df_list=  [movie_gross,imdb_title_basics, imdb_ratings]
df_list_names = ['movie_gross','imdb_title_basic','imdb_ratings']
for i, n in zip(df_list, df_list_names):
    print(f'{n}:')
    print(i.isna().sum(), '\n ----\n') 

movie_gross:
title                0
studio               5
domestic_gross      28
foreign_gross     1350
year                 0
dtype: int64 
 ----

imdb_title_basic:
tconst                 0
primary_title          0
original_title        21
start_year             0
runtime_minutes    31739
genres              5408
dtype: int64 
 ----

imdb_ratings:
tconst           0
averagerating    0
numvotes         0
dtype: int64 
 ----



All datasets contain missing values except the `imdb_rating` . This means we will deal with the missing values of the rest datasets.

For `Movie_gross` dataset , we can conclude movie entries with null values did not have domestic gross neither foreign income for that year, hence replacing the value with 0. This occurs for `domestic_gross` and `foreign_gross` .

Studio column contains 5 missing values, we replace this with `unknown`

In [55]:
#replace the missing values of foreign_gross and domestic_gross with 0.
movie_gross['domestic_gross'] = movie_gross.domestic_gross.fillna(0.0)  
movie_gross['foreign_gross'] = movie_gross.foreign_gross.fillna(0.0) 

#replace missing values of studio with 'Unknown'
movie_gross['studio'] = movie_gross.studio.fillna('Unknown')

Lastly, for `imdb_title_df` , we have 21 missing values in the original_title.<br>
 - The values in `original_title` to be replaced with `primary_title` 
- runtime_minutes are vital for the analysis , the missing values to be replaced with median value of the column.
- Missing values in the genres column will be dropped; genre of each entry is very crucial for analysis .

In [57]:
#replace missing values in the original_title column with the primary title
imdb_title_basics['original_title'] = imdb_title_basics['original_title'].fillna(imdb_title_basics['primary_title'])

# replace  missing values of the runtime_minutes with the median value
imdb_title_basics['runtime_minutes'] = imdb_title_basics['runtime_minutes'].fillna(imdb_title_basics['runtime_minutes'].median())

#drop the missing values in the genres column
imdb_title_basics = imdb_title_basics.dropna(subset=['genres'])

###  Check for duplicate values

In [58]:

#check duplicate records for imdb_ratings dataframe
imdb_ratings[imdb_ratings.duplicated(keep=False, subset=['tconst'])].sort_values(by='tconst')

Empty DataFrame
Columns: [tconst, averagerating, numvotes]
Index: []

From the above output,Imdb_ratings doesn't have any duplicates

In [59]:
#display all duplicate records for movie gross 'title' column
movie_gross[movie_gross.duplicated(keep=False, subset='title')].sort_values(by='title')

title  studio  domestic_gross foreign_gross  year
317   Bluebeard  Strand         33500.0          5200  2010
3045  Bluebeard   WGUSA         43100.0             0  2017

By further analysing this entries, each stands out to be a differnt movie and not duplicate values. Different year, studios and gross amount.

In [63]:
#display all duplicate records for imdb_title_basics dataframe using 'primary_title','start_year' and 'genres' columns
imdb_title_basics[imdb_title_basics.duplicated(keep=False, subset=['primary_title','start_year'])].sort_values(by=['primary_title','start_year'])[25:50]

tconst         primary_title        original_title  start_year  \
51820   tt3072686              5 People              5 People        2013   
55539   tt3252324              5 People              5 People        2013   
46602   tt2771990                7 Days                7 Days        2015   
88065   tt5112592                7 Days                Nanoka        2015   
104504  tt6111436                    88                    88        2016   
105408  tt6160528                    88                    88        2016   
86522   tt5017134                  8:30                  8:30        2017   
114899  tt6804666                  8:30                  8:30        2017   
31052   tt2180277            90 Minutes           90 minutter        2012   
51474   tt3054786            90 Minutes            90 Minutes        2012   
115249  tt6834356                    96                    96        2018   
117657  tt7019842                    96                    96        2018   
104657  tt6119224  A Ballad of Maladies  A Ballad of Maladies        2016   
105960  tt6200522  A Ballad of Maladies  A Ballad of Maladies        2016   
9344    tt1554091         A Better Life         A Better Life        2011   
25786   tt2027265         A Better Life     Une vie meilleure        2011   
76703   tt4387314     A Christmas Carol     A Christmas Carol        2018   
131996  tt8232648     A Christmas Carol     A Christmas Carol        2018   
142558  tt9387380     A Christmas Carol     A Christmas Carol        2018   
142749  tt9416854     A Christmas Carol     A Christmas Carol        2018   
132745  tt8308496    A Coach's Daughter         Córka trenera        2018   
143706  tt9570418    A Coach's Daughter    A Coach's Daughter        2018   
73546   tt4191350           A Courtship           A Courtship        2015   
88715   tt5136180           A Courtship           A Courtship        2015   
101588  tt5936866          A Dark Place         Steel Country        2018   

        runtime_minutes                     genres  
51820              87.0                   Thriller  
55539              87.0      Comedy,Drama,Thriller  
46602              87.0                     Sci-Fi  
88065             110.0                      Drama  
104504            102.0                Documentary  
105408            103.0                Documentary  
86522              70.0                    Mystery  
114899             70.0                    Mystery  
31052              88.0                      Drama  
51474              94.0                   Thriller  
115249             87.0     Crime,Mystery,Thriller  
117657            158.0              Drama,Romance  
104657             86.0          Documentary,Music  
105960             86.0  Documentary,History,Music  
9344               98.0              Drama,Romance  
25786             110.0                      Drama  
76703              87.0                    Fantasy  
131996             81.0                      Drama  
142558            121.0       Drama,Family,Musical  
142749             51.0                  Animation  
132745             93.0                      Drama  
143706             95.0                      Drama  
73546              71.0        Documentary,Romance  
88715              71.0                Documentary  
101588             89.0           Mystery,Thriller

`imdb_title_basics` has duplicate records based on the `primary_title` and the `start_year`. Some of the duplicate records have slightly different genres while others have missing values in either the runtime_minutes or genre columns. <br>
The duplicate records will be dropped, while keeping the entry with the least number of missing values.

In [65]:
#assign a different variable name to a copy of the imdb_title_basics dataframe
imdb_title_df = imdb_title_basics.copy()

#assign a new column, 'missing', to the sum of missing values in each row
#sort the dataframe by values of the 'primary_title' and 'missing'
imdb_title_df = imdb_title_df.assign(missing=imdb_title_df.isnull().sum(axis=1)).sort_values(by=['primary_title', 'missing'], ascending=True)

# preview of duplicated records
imdb_title_df[imdb_title_df.duplicated(keep=False, subset=['primary_title','start_year'])][:20]

tconst             primary_title            original_title  \
103890  tt6085916                 (aguirre)                 (aguirre)   
106201  tt6214664                 (aguirre)                 (aguirre)   
129962  tt8032828  100 Milioni di bracciate  100 Milioni di bracciate   
129979  tt8034014  100 Milioni di bracciate  100 Milioni di bracciate   
20394   tt1855110                       180                       180   
24194   tt1979163                       180                       180   
65431   tt3735338                      1989                      1989   
72121   tt4119270                      1989                      1989   
50212   tt2979366                   21 Days                   21 Days   
75495   tt4312624                   21 Days                    21 den   
86006   tt4981966                        24                        24   
98602   tt5765582                        24                        24   
29202   tt2123129                         3                         3   
38370   tt2379653                         3                         3   
96327   tt5618160                         3                         3   
135287  tt8558372                         3                         3   
144337  tt9653930                   3. Elma                   3. Elma   
144392  tt9660588                   3. Elma                   3. Elma   
145118  tt9773302                   3. Elma                   3. Elma   
59640   tt3463230                        37                        37   

        start_year  runtime_minutes                        genres  missing  
103890        2016             97.0         Biography,Documentary        0  
106201        2016             98.0  Biography,Comedy,Documentary        0  
129962        2017             87.0                     Biography        0  
129979        2017             87.0                     Biography        0  
20394         2011            121.0                 Drama,Romance        0  
24194         2011            107.0                  Comedy,Drama        0  
65431         2014             54.0                   Documentary        0  
72121         2014             97.0  Documentary,History,Thriller        0  
50212         2014             89.0               Horror,Thriller        0  
75495         2014             70.0                   Documentary        0  
86006         2016            164.0        Action,Sci-Fi,Thriller        0  
98602         2016             87.0                      Thriller        0  
29202         2012            145.0         Drama,Mystery,Romance        0  
38370         2012            115.0                  Comedy,Drama        0  
96327         2018             80.0       Horror,Mystery,Thriller        0  
135287        2018             17.0                   Documentary        0  
144337        2014             65.0                         Drama        0  
144392        2014             65.0                         Drama        0  
145118        2014             65.0                         Drama        0  
59640         2018             90.0                         Drama        0

Then, we drop duplicates to obatin values entries with least missing values

In [67]:
imdb_title_df = imdb_title_df.drop_duplicates(keep='first', subset=['primary_title','start_year']).drop(columns='missing')

To check for any more duplicates:

In [68]:
#recheck for duplicates
imdb_title_df[imdb_title_df.duplicated(keep=False, subset=['primary_title','start_year'])].sort_values(by=['primary_title','start_year'])

Empty DataFrame
Columns: [tconst, primary_title, original_title, start_year, runtime_minutes, genres]
Index: []

### Amend data types

Each column should have correspondingly valid data types so that analsysis can occur withou any hitches. The function `df_inspect` contains a method `.info` that provided information for each dataset paased through. <br>

We can find the dataype for each column using `.dtypes` . As below :

In [69]:
#check the data type of each column 
print(f'movie_gross data types: \n {movie_gross.dtypes} \n ---')
print(f'\n imdb_title_df data types: \n {imdb_title_df.dtypes} \n ---')
print(f'\n imdb_ratings data types: \n {imdb_ratings.dtypes} \n ---')

movie_gross data types: 
 title              object
studio             object
domestic_gross    float64
foreign_gross      object
year                int64
dtype: object 
 ---

 imdb_title_df data types: 
 tconst              object
primary_title       object
original_title      object
start_year           int64
runtime_minutes    float64
genres              object
dtype: object 
 ---

 imdb_ratings data types: 
 tconst            object
averagerating    float64
numvotes           int64
dtype: object 
 ---


Notice that column foreign_gross has object values instead of float values to represent number values. So we convert that by :


In [70]:
#Convert the data type of the foreign_gross column from object to float
movie_gross['foreign_gross'] = movie_gross['foreign_gross'].str.replace(",","").apply(float)

We call out the type of `foreign_gross` to confirm :

In [71]:
movie_gross['foreign_gross'].dtype

dtype('float64')

### Amend error values

In the Boc office dataset, we notice there are movie that have pre-planned release year(under `start_year` column). So, we align data upto year of current record(2022):

In [72]:
#check for years greater than 2022 in imdb_title_df
imdb_title_df[imdb_title_df['start_year'] > 2022]

tconst                              primary_title  \
89506    tt5174640                                  100 Years   
16337    tt1757678                                   Avatar 3   
52213    tt3095356                                   Avatar 4   
96592    tt5637536                                   Avatar 5   
105187   tt6149054  Fantastic Beasts and Where to Find Them 5   
2483    tt10255736                    Untitled Marvel Project   
2949    tt10300398                    Untitled Star Wars Film   
106865   tt6258542         Wraith of the Umbra and Eidolon II   

                                   original_title  start_year  \
89506                                   100 Years        2115   
16337                                    Avatar 3        2023   
52213                                    Avatar 4        2025   
96592                                    Avatar 5        2027   
105187  Fantastic Beasts and Where to Find Them 5        2024   
2483                      Untitled Marvel Project        2023   
2949                      Untitled Star Wars Film        2026   
106865         Wraith of the Umbra and Eidolon II        2023   

        runtime_minutes                    genres  
89506              87.0                     Drama  
16337              87.0    Action,Adventure,Drama  
52213              87.0  Action,Adventure,Fantasy  
96592              87.0  Action,Adventure,Fantasy  
105187             87.0  Adventure,Family,Fantasy  
2483               87.0                    Action  
2949               87.0                   Fantasy  
106865             87.0   Adventure,Drama,Fantasy

In [73]:
# drop records with start_year more than 2022
imdb_title_df = imdb_title_df[imdb_title_df['start_year'] <= 2022]

In [74]:
#recheck
imdb_title_df[imdb_title_df['start_year'] > 2022]

Empty DataFrame
Columns: [tconst, primary_title, original_title, start_year, runtime_minutes, genres]
Index: []

In [75]:
#check for years greater than 2022 in movie_gross
movie_gross[movie_gross['year'] > 2022]

Empty DataFrame
Columns: [title, studio, domestic_gross, foreign_gross, year]
Index: []

And we are done with cleaning the data.

## Merging of dataframes

I merge the imdb_title_df with imdb_rating to provide more entries within a row

In [76]:
#merge imdb_title_df and imdb_ratings using an inner join
joined_imdb_df = imdb_title_df.merge(imdb_ratings, on='tconst', how='inner')
joined_imdb_df.head(10)

tconst          primary_title        original_title  start_year  \
0  tt1699720  !Women Art Revolution  Women Art Revolution        2010   
1  tt2346170       #1 Serial Killer      #1 Serial Killer        2013   
2  tt3120962                     #5                    #5        2013   
3  tt5255986                    #66                   #66        2015   
4  tt7853996           #ALLMYMOVIES          #ALLMYMOVIES        2015   
5  tt9844890              #AbroHilo             #AbroHilo        2019   
6  tt6170868                  #BKKY                 #BKKY        2016   
7  tt5074174     #BeRobin the Movie    #BeRobin the Movie        2015   
8  tt4353986                #Beings               #Beings        2015   
9  tt6856592              #Captured             #Captured        2017   

   runtime_minutes                    genres  averagerating  numvotes  
0             83.0               Documentary            6.9       196  
1             87.0                    Horror            5.6        40  
2             68.0  Biography,Comedy,Fantasy            6.8         6  
3            116.0                    Action            5.3        18  
4             87.0               Documentary            8.3        21  
5             52.0               Documentary            6.6         8  
6             75.0                     Drama            7.4        23  
7             41.0               Documentary            9.0        31  
8             56.0                  Thriller            5.3        18  
9             81.0                  Thriller            2.8       212